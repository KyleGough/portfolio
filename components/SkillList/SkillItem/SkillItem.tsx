import { useObserveElement } from '@hooks/useObserveElement';
import { clsx } from 'clsx';
import Image from 'next/image';
import React, { useEffect, useState } from 'react';

interface SkillItemProps {
  type: string;
  progress: number;
  progressClass: string;
  name: string;
  description: string;
  logo: string;
}

export const SkillItem: React.FC<SkillItemProps> = ({
  type,
  progress,
  progressClass,
  name,
  description,
  logo,
}) => {
  const [progressRef, isVisible] = useObserveElement<HTMLDivElement>({
    threshold: 0,
  });
  const [animatedIn, setAnimatedIn] = useState(false);
  const durationVar = {
    '--duration-factor': progress,
  } as React.CSSProperties;

  useEffect(() => {
    if (!animatedIn && isVisible) {
      (progressRef.current as HTMLDivElement).classList.replace(
        '-translate-x-full',
        'translate-x-0'
      );
      setAnimatedIn(true);
    }
  }, [progressRef, isVisible, animatedIn]);

  return (
    <div className="flex flex-row justify-center py-6">
      <div className="w-3/12 lg:w-4/12 flex items-center">
        <div className={clsx(type, 'progress overflow-hidden')}>
          <div
            ref={progressRef}
            className={clsx(
              progressClass,
              'custom-duration progress-bar',
              'transition-transform ease-in-out -translate-x-full'
            )}
            style={durationVar}
          ></div>
        </div>
      </div>
      <div className="w-9/12 lg:w-8/12 ml-6 flex items-center">
        <div
          className={clsx(
            { 'opacity-100': animatedIn },
            { 'opacity-0': !animatedIn },
            'mr-4 w-8 h-8 transition-opacity duration-500 relative'
          )}
        >
          <Image src={logo} alt={`${name} logo`} layout="fill" />
        </div>
        <div>
          <p>
            <span className="font-extrabold text-lg">{name}</span> -{' '}
            <span className="text-sm">{description}</span>
          </p>
        </div>
      </div>
    </div>
  );
};
