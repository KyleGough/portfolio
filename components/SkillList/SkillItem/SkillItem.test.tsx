import { mockIntersectionObserver } from '@mocks/mockIntersectionObserver';
import { render, screen } from '@testing-library/react';
import React from 'react';

import { SkillItem } from './SkillItem';

const skill = {
  type: 'progress-web',
  progress: 75,
  progressClass: 'w-[75%]',
  name: 'React',
  description: 'Portfolio, LucidLab, Minesweeper',
  logo: '/#logo',
};

describe('SkillItem component', () => {
  beforeEach(mockIntersectionObserver);

  it('renders', async () => {
    const { container } = render(<SkillItem {...skill} />);

    expect(screen.getByText('React')).toBeVisible();
    expect(screen.getByText('Portfolio, LucidLab, Minesweeper')).toBeVisible();
    expect(container).toMatchSnapshot();
  });
});
