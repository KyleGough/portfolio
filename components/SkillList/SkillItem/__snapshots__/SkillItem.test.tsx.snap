// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkillItem component renders 1`] = `
<div>
  <li
    class="flex flex-row py-6 ml-8 items-center relative before:absolute before:h-2 before:w-2 before:rounded-full before:-left-6"
  >
    <div
      class="mr-4 w-8 h-8 duration-500 relative opacity-0 transition-opacity duration-1000"
    >
      <img
        alt="React logo"
        data-nimg="1"
        decoding="async"
        height="30"
        loading="lazy"
        src="/_next/image?url=%2F%23logo&w=64&q=75"
        srcset="/_next/image?url=%2F%23logo&w=32&q=75 1x, /_next/image?url=%2F%23logo&w=64&q=75 2x"
        style="color: transparent;"
        width="30"
      />
    </div>
    <div>
      <p>
        <span
          class="font-extrabold text-lg"
        >
          React
        </span>
         -
         
        <span
          class="text-sm"
        >
          Portfolio, LucidLab, Minesweeper
        </span>
      </p>
    </div>
  </li>
</div>
`;
